api/jwt.go

```diff
@@ -3,77 +3,227 @@ package api
-	}
-	return token, nil
-}
-
 // ValidateToken verifies the validity of a JWT token, both in terms of validity
 // of the signature and expiration time.
+//
+// Example token:
+//
+// Header:
+//
+//{
+//  "alg": "RS256",
+//  "kid": "a2aa9739-d753-4a0d-87ee-61f101050277",
+//  "typ": "JWT"
+//}
+//
+// Payload:
+//
+//{
+//  "exp": 1607594172,
+//  "iat": 1607593272,
+//  "iss": "https://siasky.xyz/",
+//  "jti": "1e5872ae-71d8-49ec-a550-4fc6163cbbf2",
+//  "nbf": 1607593272,
+//  "session": {
+//    "active": true,
+//    "authenticated_at": "2020-12-09T16:09:35.004003Z",
+//    "expires_at": "2020-12-10T16:09:35.004003Z",
+//    "id": "9911ad26-e47f-4ec4-86a1-fbbc7fd5073e",
+//    "identity": {
+//      "id": "695725d4-a345-4e68-919a-7395cb68484c",
+//      "recovery_addresses": [
+//        {
+//          "id": "e2d847e1-1885-4edf-bccb-64b527b30096",
+//          "value": "ivaylo@nebulous.tech",
+//          "via": "email"
+//        }
+//      ],
+//      "schema_id": "default",
+//      "schema_url": "https://siasky.xyz/secure/.ory/kratos/public/schemas/default",
+//      "traits": {
+//        "email": "ivaylo@nebulous.tech",
+//        "name": {
+//          "first": "Ivaylo",
+//          "last": "Novakov"
+//        }
+//      },
+//      "verifiable_addresses": [
+//        {
+//          "id": "953b0c1a-def9-4fa2-af23-fb36c00768d2",
+//          "status": "pending",
+//          "value": "ivaylo@nebulous.tech",
+//          "verified": false,
+//          "verified_at": null,
+//          "via": "email"
+//        }
+//      ]
+//    },
+//    "issued_at": "2020-12-09T16:09:35.004042Z"
+//  },
+//  "sub": "695725d4-a345-4e68-919a-7395cb68484c"
+//}
 func ValidateToken(t string) (*jwt.Token, error) {
-	token, err := jwt.Parse(t, func(token *jwt.Token) (interface{}, error) {
-		//Make sure that the token method conform to "SigningMethodHMAC"
-		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
-			return nil, errors.New(fmt.Sprintf("unexpected signing method: %v", token.Header["alg"]))
-		}
-		return []byte(os.Getenv("ACCESS_SECRET")), nil
-	})
+	token, err := jwt.Parse(t, keyForToken)
 	if err != nil {
 		return nil, err
 	}
 	if !token.Valid {
 		return nil, errors.New("token is invalid")
 	}
+	// TODO Verify issuer, scope, etc.?
 	return token, nil
 }
 
-// extractToken extracts the JWT token from the request and returns it. Returns
-// an empty string if there is no token.
-func extractToken(r *http.Request) string {
+// keyForToken is a helper function that finds a suitable key for validating the
+// given token among the public keys provided by Oathkeeper.
+func keyForToken(token *jwt.Token) (interface{}, error) {
+	if _, ok := token.Method.(*jwt.SigningMethodRSA); !ok {
+		return nil, errors.New(fmt.Sprintf("unexpected signing method: %v", token.Header["alg"]))
+	}
+	keySet, err := oathkeeperPublicKeys()
+	if err != nil {
+		return nil, err
+	}
+	// Should this be replaced by a deferred recover in case the value is not
```

Comment:
I would mark open questions like these as `TODO`.

Reply:
To answer the question, I'm kind of surprised that a `Header` can have anything other than a string. Usually it's a string dictionary right? I guess it won't be much of an issue, aren't we only providing JWTs here that are given out by ORYs oathkeeper service?